### PR TITLE
Improve fast pandas and flexABLE startegies for better performance

### PR DESCRIPTION
**Pull request by nick-harder**
# Pull Request

## Related Issue
Closes #[issue number] (if applicable)

## Description
After the latest commit with the FastSeries I have noticed a huge slowßdown for long and large simulations. Here I took the effort to find the reasons and came upon multiple places. At the end the problem is now fixed and the speed up for large simulations compared to current main is 4x. For small simulations for one year (example_01a) the speed up is 2x. Still some further testing is needed. This results in a total speed up of 10x compared to the last stable release fr large simulations and 8x for small simulations.

## Changes Proposed
-change how average operation time and down time is calculated
-implement static methods for loc and at methods in fast series
-optimize FastIndex to remove overhead of the idx_from_date method
-do not create FastIndex constantly, instead use the slice
-for demand unit, compute the volume for the specific index instead of the whole index
-optimize write_dispatch and write_market_dispatch in the output role to collect the values in a buffer and process them at once resulting in a 2x speed up
-optimize separate_orders function to improve performance
-revise calculate_reward functions for flexable startegies to remove unrequired parameters and calculations
-remove save_frequency_hours: null from all configs and replace with 720 (one month) to avoid confusion for users

## Testing
[Describe the testing you've done, including any specific test cases or scenarios]

## Checklist
Please check all applicable items:

- [x] Code changes are sufficiently documented (docstrings, inline comments, `doc` folder updates)
- [ ] New unit tests added for new features or bug fixes
- [x] Existing tests pass with the changes
- [x] Reinforcement learning examples are operational (for DRL-related changes)
- [x] Code tested with both local and Docker databases
- [x] Code follows project style guidelines and best practices
- [ ] Changes are backwards compatible, or deprecation notices added
- [ ] New dependencies added to `pyproject.toml`
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included
- [x] Consent to release this PR's code under the GNU Affero General Public License v3.0

## Additional Notes (if applicable)
[Any additional information, concerns, or areas you want reviewers to focus on]

## Screenshots (if applicable)
[Add screenshots to demonstrate visual changes]

### CODE REVIEW

**Review by maurerle — COMMENTED** [review]
I left some remarks.

I would change some things in the output role, so that we do not need to bookkeep two types of buffers.
Makes sense to move the pandas conversion into the writing, quite interesting that this has such a large impact.

File: assume/strategies/flexable.py [hunk 1]
```diff
@@ -441,6 +496,9 @@ def calculate_EOM_price_if_off(
         float: The inflexible bid price of the unit.
 
     """
+    # ensure that we don't divide by 0
+    avg_op_time = max(avg_op_time, 1)
```
[the remark this answers is not in the document]
Comment (nick-harder) [reply assume/strategies/flexable.py]: fixed

File: assume/common/base.py [hunk 2]
```diff
@@ -139,24 +143,54 @@ def set_dispatch_plan(
...
+                if was_operating and not now_operating:  # Transition to downtime
+                    total_op_time -= 1
+                    total_down_time += 1
+                elif not was_operating and now_operating:  # Transition to operating
+                    total_op_time += 1
+                    total_down_time -= 1
+
+        # Recalculate averages
+        self.avg_op_time = total_op_time / total_periods
+        self.avg_down_time = total_down_time / total_periods
```
[the remark this answers is not in the document]
Comment (nick-harder) [reply assume/common/base.py:190]: @maurerle get_operation_time and get_average_operation_times are two different functions. The first one checks if the unit can be turned on or off at this moment, and here the min down and up times are taken into account. The second one is used to bring the start up costs into the bid. So each bid contains only one part of the start up cost, which is the cost devided by the avg_op_time. old behavior was calculating this values always from index[0] and till the end. new behavior updates the values as we go

File: assume/common/units_operator.py [hunk 3]
```diff
@@ -60,7 +60,7 @@ def __init__(
 
         self.available_markets = available_markets
         self.registered_markets: dict[str, MarketConfig] = {}
-        self.last_sent_dispatch = defaultdict(lambda: 0)
+        self.last_sent_dispatch = defaultdict(lambda: datetime(1970, 1, 1, 1, 0))
```
[the remark this answers is not in the document]
Comment (nick-harder) [reply assume/common/units_operator.py:63]: @maurerle the problem is, that when the simulation starts, the units use this 0 as initial time step using it as start and try to access values in FastSeries with index (1970,1,1,1,0), which doesn't exist and not in the index. we need a better solution for this problem. this start of the simulation from 1970 is really problematic in many cases

[the remark this answers is not in the document]
Comment (nick-harder) [reply assume/common/units_operator.py]: restored to the original values

File: assume/common/base.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (maurerle) [reply assume/common/base.py:191]: So if I have a plant which was never turned on (self.avg_op_time = 1), then I have to pay the whole startup cost for the next hour at once.
The next hour, the plant is already running:

````
if was_operating and not now_operating:  # Transition to downtime
    total_op_time -= 1
    total_down_time += 1
elif not was_operating and now_operating:  # Transition to operating
    total_op_time += 1
    total_down_time -= 1
```

so nothing happens, and self.avg_op_time stays at simulation_length/simulation_length == 1.

If we turn the plant of, we run `total_op_time -= 1` - so it is `(simulation_length-1)/simulation_length` until it is turned on again.

This does not look correct to me..?

File: tests/test_utils.py
[the remark this answers is not in the document]
Comment (nick-harder) [reply tests/test_utils.py:1]: I find it hard to always say exactly by how much it should be faster or not. On one system it is 5x, on another system it is different. It is not easy to replicate these speed up tests. Maybe you have another idea, which wouldn't require adjusting the tests always?

**Comment by nick-harder** [comment]
will create a new branch with more focused changes

**Pull request closed without merging**